README.md
# AminoX 

http://github.com/tendermint/aminox

AminoX is an encoding/decoding library for structures.

In AminoX, all structures are declared in a restricted set of Go.

From those declarations, AminoX generates Protobuf3 compatible binary bytes.

For faster encoding and decoding, AminoX also generates Protobuf3 schemas, and
also generates translation logic between the original Go declarations and
Protobuf3/protoc generated Go.  In this way, AminoX supports a restricted set
of Protobuf3.

Though AminoX supports a subset of Protobuf3 and uses it to optimize encoding
and decoding, it is NOT intended to be a Protobuf3 library -- complete support
of Protobuf3 is explicitly not its design goal.

You can see the performance characteristics of the improvements in [XXX -- this
exists, but i forget the filename.].

The gist is that in recent versions of Go, the reflection-based binary
encoding/decoding system is about 3x slower than the protoc generated Go ones,
and that the translation system works pretty well, accounting for only ~25% of
the total time, so the performance hit isn't fatal.

While creating and finalizing this library, which I believe it is, roughly, the
final form of a well structured piece of software according to my tastes, it
occurred to me that Proto3 is a complexification that is not needed in well
written software, and that it mostly serves systems that are part of the mass
public surveillance database amassed in datacenters across the world.  Somewhere,
there is a proto3 field number and value that describes some new aspect about me,
in several instances of Google's massive database. 

What I want instead is a language, and for that, the original implementation
of Amino is better suited.

# AminoX JSON

This is experimental and subject to change.

## Amino in the Wild

* Amino:binary spec in [Tendermint](
https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/encoding.md)


# Amino Spec

#### Registering types and packages

Previous versions of Amino used to require a local codec where types must be
registered.  With the change to support Any and type URL strings,
we no longer need to keep track of local codecs, unless we want to override
default behavior from global registrations.

Each package should declare in a package local file (by convention called amino.go)
which should look like the following:

```go
// see github.com/tendermint/go-amino-x/protogen/example/main.go
package main

import (
	"github.com/tendermint/go-amino-x"
	"github.com/tendermint/go-amino-x/genproto/example/submodule"
)

var Package = amino.RegisterPackage(
	amino.NewPackage(
		"main", // The Go package path
		"main", // The (shorter) Proto3 package path (no slashes).
		amino.GetCallersDirname(),
	).WithDependencies(
		submodule.Package, // Dependencies must be declared (for now).
	).WithTypes(
		StructA{}, // Declaration of all structs to be managed by Amino.
		StructB{}, // For example, protogen to generate proto3 schema files.
		&StructC{}, // If pointer receivers are preferred when decoding to interfaces.
	),
)
```

You can still override global registrations with local \*amino.Codec state.
This is used by genproto.P3Context, which may help development while writing
migration scripts.  Feedback welcome in the issues section.

## Unsupported types

### Floating points
Floating point number types are discouraged as [they are generally
non-deterministic](http://gafferongames.com/networking-for-game-programmers/floating-point-determinism/).
If you need to use them, use the field tag `amino:"unsafe"`.

### Enums
Enum types are not supported in all languages, and they're simple enough to
model as integers anyways.

### Maps
Maps are not currently supported.  There is an unstable experimental support
for maps for the Amino:JSON codec, but it shouldn't be relied on.  Ideally,
each Amino library should decode maps as a List of key-value structs (in the
case of langauges without generics, the library should maybe provide a custom
Map implementation).  TODO specify the standard for key-value items.

## Amino and Proto3

Amino objects are a subset of Proto3.
* Enums are not supported.
* Nested message declarations are not supported.

Amino extends Proto3's Any system with a particular concrete type
identification format (disfix bytes).

## Amino and Go 

Amino objects are a subset of Go.
* Multi-dimensional slices/arrays are not (yet) supported.
* Floats are nondeterministic, so aren't supported by default.
* Complex types are not (yet) supported.
* Chans, funcs, and maps are not supported.
* Pointers are automatically supported in go-amino but it is an extension of
  the theoretical Amino spec.

Amino, unlike Gob, is beyond the Go language, though the initial implementation
and thus the specification happens to be in Go (for now).

## Limitations

* Pointer types in arrays and slices lose pointer information.
* Nested pointers are not allowed.
* Recursive ReprType not allowed.


CHANGELOG.md
# Changelog

## Go-Amino-X 1.0.0-rc0 (June 28th, 2020)

BREAKING CHANGE:
 - Many changes have been made; go-amino-x binary and JSON are not backwards
   compatible.
 - Interfaces are encoded as google.protobuf.Any. As representations are
   deterministic, local Codecs are no longer required.  Just register packages
at the global level.
 - Nested pointers are not allowed.  Pointers to primitives, structs, and
   interfaces are allowed.  See
https://github.com/tendermint/go-amino-x/pull/290#issuecomment-650717598.
 - IMPORTNT: Field tag for lists "EmptyElements" replaced with "NilElements",
   with opposite behavior.
 - MarshalBinaryBare() renamed to Marshal().  The default encoding is bare,
use length-prefixed for streaming messages.
 - [Un]Marshal...LengthPrefixed... -> ...Sized...

NEW FEATURES:
 - DeepEquals() placeholder, using binary encoding/decoding for equality testing (for now).

## 0.15.0 (May 2, 2018)

BREAKING CHANGE:
 - JSON: Skip null fields on decoding when the `json:"omitempty"` tag is set. 
 This matches the encoding behaviour ([#261]).  
 
IMPROVEMENTS:
 - Amino becomes a go-module (requires go 1.11) but keeps dep support for backwards compatibility ([#255]). 

 
[#255]: https://github.com/tendermint/go-amino-x/pull/255 
[#261]: https://github.com/tendermint/go-amino-x/issues/261
 
## 0.14.1 (November 6, 2018)

IMPROVEMENTS:
 - go-amino compiles again on 32-bit platforms ([#242])
 
[#242]: https://github.com/tendermint/go-amino-x/pull/242

## 0.14.0 (October 26, 2018)

BREAKING CHANGE:
 - Switch default encoding of unsigned ints (`int`, `int32`, `int64`) to be on par with [proto3's] variable length 
 encoding (of `int32`, `int64`) ([#237])
 
 [proto3's]: https://developers.google.com/protocol-buffers/docs/proto#scalar
 [#237]: https://github.com/tendermint/go-amino-x/issues/237

## 0.13.0 (October 15, 2018)

BREAKING CHANGE:
 - `time.Time`: validate seconds since 1970 are in valid range during encoding; i.e. in the interval [-62135596800, 253402300800)
 - `time.Time`: match encoding of time.Time to protobuf's well known type [Timestamp] ([#224])
 - Rename `Marshal` to `MarshalSized` ([#222])
 
[Timestamp]: https://github.com/protocolbuffers/protobuf/blob/d2980062c859649523d5fd51d6b55ab310e47482/src/google/protobuf/timestamp.proto#L123-L135

IMPROVEMENTS:
  - Add `MustUnmarshalJSON` and `MustMarshalJSON` that panic if an error occurs ([#228])
  - Optimize performance by removing `defer`s in `getTypeInfo` ([#227] by @ValarDragon) 
  - Optimize performance by removing unnecessary allocation in `UvarintSize` ([#225] by [@ValarDragon])
  - Add test that shows compatibility with stdlib (JSON time encoding) ([#131] by [@odeke-em]) 
  
[#224]: https://github.com/tendermint/go-amino-x/pull/224
[#222]: https://github.com/tendermint/go-amino-x/pull/222
[#228]: https://github.com/tendermint/go-amino-x/pull/228
[#227]: https://github.com/tendermint/go-amino-x/pull/227
[#225]: https://github.com/tendermint/go-amino-x/pull/225
[#131]: https://github.com/tendermint/go-amino-x/pull/131
[@ValarDragon]: https://github.com/ValarDragon
[@odeke-em]: https://github.com/odeke-em

## 0.12.0 (August 4, 2018)

BREAKING CHANGE:
 - Write empty (non-nil) struct pointers, unless (is list element and empty_elements isn't set) #206

## 0.11.1 (July 17, 2018)

IMPROVEMENTS:
 - Remove dependency on tmlibs/common

## 0.11.0 (June 19, 2018)

BREAKING CHANGE:

 - Do not encode zero values in `EncodeTime`
 (to match proto3's behaviour) (#178, #190)
 - Do not encode empty structs, unless explicitly enforced
 via `amino:"write_empty"` (to match proto3's behaviour) (#179)

IMPROVEMENTS:
 - DecodeInt{8, 16} negative limit checks (#125)

## 0.10.1 (June 15, 2018)

FEATURE:

 - [aminoscan] aminoscan --color will print ASCII bytes in different colors

BUG FIXES:
 - do not err if prefix bytes are exactly 4 (for registered types)

## 0.10.0 (June 12, 2018)

BREAKING CHANGE:

 - 100% Proto3 compatibility for primitive types, repeated fields, and embedded structs/messages.
 - BigEndian -> LittleEndian
 - [u]int[64/32] is (signed) Varint by default, "fixed32" and "fixed64" to use 4 and 8 byte types.
 - Amino:JSON [u]int64 and ints are strings.
 - Enforce UTC timezone for JSON encoding of time.

## 0.9.11 (May 27, 2018)

NEW FEATURES:

 - Seal() on a codec to prevent further modifications. #150
 - Global Marshal/Unmarshal methods on a sealed codec with nothing registered.

## 0.9.10 (May 10, 2018)

BREAKING CHANGE:

 - Amino:JSON encoding of interfaces use the registered concrete type name, not the disfix bytes.

## 0.9.9 (May 1, 2018)

BUG FIXES:

 - MarshalAmino/UnmarshalAmino actually works (sorry!)


amino.go
package amino

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"path"
	"reflect"
	"runtime"
	"time"

	"github.com/pkg/errors"
	"github.com/tendermint/go-amino-x/pkg"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/durationpb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// Package "pkg" exists So dependencies can create Packages.
// We export it here so this amino package can use it natively.
type Package = pkg.Package
type Type = pkg.Type

var (
	// Global methods for global auto-sealing codec.
	gcdc *Codec

	// we use this time to init. an empty value (opposed to reflect.Zero which gives time.Time{} / 01-01-01 00:00:00)
	emptyTime time.Time

	// ErrNoPointer is thrown when you call a method that expects a pointer, e.g. Unmarshal
	ErrNoPointer = errors.New("expected a pointer")
)

const (
	unixEpochStr = "1970-01-01 00:00:00 +0000 UTC"
	epochFmt     = "2006-01-02 15:04:05 +0000 UTC"
)

func init() {
	gcdc = NewCodec().Autoseal()
	var err error
	emptyTime, err = time.Parse(epochFmt, unixEpochStr)
	if err != nil {
		panic("couldn't parse empty value for time")
	}
}

func MarshalSized(o interface{}) ([]byte, error) {
	return gcdc.MarshalSized(o)
}

func MarshalSizedWriter(w io.Writer, o interface{}) (n int64, err error) {
	return gcdc.MarshalSizedWriter(w, o)
}

func MustMarshalSized(o interface{}) []byte {
	return gcdc.MustMarshalSized(o)
}

func MarshalAnySized(o interface{}) ([]byte, error) {
	return gcdc.MarshalAnySized(o)
}

func MustMarshalAnySized(o interface{}) []byte {
	return gcdc.MustMarshalAnySized(o)
}

func MarshalAnySizedWriter(w io.Writer, o interface{}) (n int64, err error) {
	return gcdc.MarshalAnySizedWriter(w, o)
}

func Marshal(o interface{}) ([]byte, error) {
	return gcdc.Marshal(o)
}

func MustMarshal(o interface{}) []byte {
	return gcdc.MustMarshal(o)
}

func MarshalAny(o interface{}) ([]byte, error) {
	return gcdc.MarshalAny(o)
}

func MustMarshalAny(o interface{}) []byte {
	return gcdc.MustMarshalAny(o)
}

func UnmarshalSized(bz []byte, ptr interface{}) error {
	return gcdc.UnmarshalSized(bz, ptr)
}

func UnmarshalSizedReader(r io.Reader, ptr interface{}, maxSize int64) (n int64, err error) {
	return gcdc.UnmarshalSizedReader(r, ptr, maxSize)
}

func MustUnmarshalSized(bz []byte, ptr interface{}) {
	gcdc.MustUnmarshalSized(bz, ptr)
}

func Unmarshal(bz []byte, ptr interface{}) error {
	return gcdc.Unmarshal(bz, ptr)
}

func MustUnmarshal(bz []byte, ptr interface{}) {
	gcdc.MustUnmarshal(bz, ptr)
}

func UnmarshalAny(typeURL string, value []byte, ptr interface{}) error {
	return gcdc.UnmarshalAny(typeURL, value, ptr)
}

func MustUnmarshalAny(typeURL string, value []byte, ptr interface{}) {
	gcdc.MustUnmarshalAny(typeURL, value, ptr)
}

func MarshalJSON(o interface{}) ([]byte, error) {
	return gcdc.MarshalJSON(o)
}

func MarshalJSONAny(o interface{}) ([]byte, error) {
	return gcdc.MarshalJSONAny(o)
}

func UnmarshalJSON(bz []byte, ptr interface{}) error {
	return gcdc.UnmarshalJSON(bz, ptr)
}

func MarshalJSONIndent(o interface{}, prefix, indent string) ([]byte, error) {
	return gcdc.MarshalJSONIndent(o, prefix, indent)
}

// XXX unstable API.
func GetTypeURL(o interface{}) string {
	return gcdc.GetTypeURL(o)
}

//----------------------------------------
// Typ3

type Typ3 uint8

const (
	// Typ3 types
	Typ3Varint     = Typ3(0)
	Typ38Byte      = Typ3(1)
	Typ3ByteLength = Typ3(2)
	//Typ3_Struct     = Typ3(3)
	//Typ3_StructTerm = Typ3(4)
	Typ34Byte = Typ3(5)
	//Typ3_List       = Typ3(6)
	//Typ3_Interface  = Typ3(7)
)

func (typ Typ3) String() string {
	switch typ {
	case Typ3Varint:
		return "(U)Varint"
	case Typ38Byte:
		return "8Byte"
	case Typ3ByteLength:
		return "ByteLength"
	//case Typ3_Struct:
	//	return "Struct"
	//case Typ3_StructTerm:
	//	return "StructTerm"
	case Typ34Byte:
		return "4Byte"
	//case Typ3_List:
	//	return "List"
	//case Typ3_Interface:
	//	return "Interface"
	default:
		return fmt.Sprintf("<Invalid Typ3 %X>", byte(typ))
	}
}

//----------------------------------------
// *Codec methods

//----------------------------------------
// Marshal* methods

// MarshalSized encodes the object o according to the Amino spec,
// but prefixed by a uvarint encoding of the object to encode.
// Use Marshal if you don't want byte-length prefixing.
//
// For consistency, MarshalSized will first dereference pointers
// before encoding.  MarshalSized will panic if o is a nil-pointer,
// or if o is invalid.
func (cdc *Codec) MarshalSized(o interface{}) ([]byte, error) {
	cdc.doAutoseal()

	// Write the bytes here.
	var buf = new(bytes.Buffer)

	// Write the bz without length-prefixing.
	bz, err := cdc.Marshal(o)
	if err != nil {
		return nil, err
	}

	// Write uvarint(len(bz)).
	err = EncodeUvarint(buf, uint64(len(bz)))
	if err != nil {
		return nil, err
	}

	// Write bz.
	_, err = buf.Write(bz)
	if err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

// MarshalSizedWriter writes the bytes as would be returned from
// MarshalSized to the writer w.
func (cdc *Codec) MarshalSizedWriter(w io.Writer, o interface{}) (n int64, err error) {
	var (
		bz []byte
		_n int
	)
	bz, err = cdc.MarshalSized(o)
	if err != nil {
		return 0, err
	}
	_n, err = w.Write(bz) // TODO: handle overflow in 32-bit systems.
	n = int64(_n)
	return
}

// Panics if error.
func (cdc *Codec) MustMarshalSized(o interface{}) []byte {
	bz, err := cdc.MarshalSized(o)
	if err != nil {
		panic(err)
	}
	return bz
}

func (cdc *Codec) MarshalAnySized(o interface{}) ([]byte, error) {
	cdc.doAutoseal()

	// Write the bytes here.
	var buf = new(bytes.Buffer)

	// Write the bz without length-prefixing.
	bz, err := cdc.MarshalAny(o)
	if err != nil {
		return nil, err
	}

	// Write uvarint(len(bz)).
	err = EncodeUvarint(buf, uint64(len(bz)))
	if err != nil {
		return nil, err
	}

	// Write bz.
	_, err = buf.Write(bz)
	if err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

func (cdc *Codec) MustMarshalAnySized(o interface{}) []byte {
	bz, err := cdc.MarshalAnySized(o)
	if err != nil {
		panic(err)
	}
	return bz
}

func (cdc *Codec) MarshalAnySizedWriter(w io.Writer, o interface{}) (n int64, err error) {
	var (
		bz []byte
		_n int
	)
	bz, err = cdc.MarshalAnySized(o)
	if err != nil {
		return 0, err
	}
	_n, err = w.Write(bz) // TODO: handle overflow in 32-bit systems.
	n = int64(_n)
	return
}

// Marshal encodes the object o according to the Amino spec.
// Marshal doesn't prefix the byte-length of the encoding,
// so the caller must handle framing.
// Type information as in google.protobuf.Any isn't included, so manually wrap
// before calling if you need to decode into an interface.
// NOTE: nil-struct-pointers have no encoding. In the context of a struct,
// the absence of a field does denote a nil-struct-pointer, but in general
// this is not the case, so unlike MarshalJSON.
func (cdc *Codec) Marshal(o interface{}) ([]byte, error) {
	cdc.doAutoseal()

	if cdc.usePBBindings {
		pbm, ok := o.(PBMessager)
		if ok {
			return cdc.MarshalPBBindings(pbm)
		} else {
			// Fall back to using relfection for native primitive types.
		}
	}

	return cdc.MarshalReflect(o)
}

// Use reflection.
func (cdc *Codec) MarshalReflect(o interface{}) ([]byte, error) {

	// Dereference value if pointer.
	var rv = reflect.ValueOf(o)
	if rv.Kind() == reflect.Ptr {
		if rv.IsNil() {
			panic("Marshal cannot marshal a nil pointer directly. Try wrapping in a struct?")
			// NOTE: You can still do so by calling
			// `.MarshalSized(struct{ *SomeType })` or so on.
		}
		rv = rv.Elem()
		if rv.Kind() == reflect.Ptr {
			panic("nested pointers not allowed")
		}
	}

	// Encode Amino:binary bytes.
	var bz []byte
	buf := new(bytes.Buffer)
	rt := rv.Type()
	info, err := cdc.getTypeInfoWLock(rt)
	if err != nil {
		return nil, err
	}
	// Implicit struct or not?
	// NOTE: similar to binary interface encoding.
	fopts := FieldOptions{}
	if !info.IsStructOrUnpacked(fopts) {
		writeEmpty := false
		// Encode with an implicit struct, with a single field with number 1.
		// The type of this implicit field determines whether any
		// length-prefixing happens after the typ3 byte.
		// The second FieldOptions is empty, because this isn't a list of
		// Typ3_ByteLength things, so however it is encoded, that option is no
		// longer needed.
		if err = cdc.writeFieldIfNotEmpty(buf, 1, info, FieldOptions{}, FieldOptions{}, rv, writeEmpty); err != nil {
			return nil, err
		}
		bz = buf.Bytes()
	} else {
		// The passed in BinFieldNum is only relevant for when the type is to
		// be encoded unpacked (elements are Typ3_ByteLength).  In that case,
		// encodeReflectBinary will repeat the field number as set here, as if
		// encoded with an implicit struct.
		err = cdc.encodeReflectBinary(buf, info, rv, FieldOptions{BinFieldNum: 1}, true, 0)
		if err != nil {
			return nil, err
		}
		bz = buf.Bytes()
	}

	// If bz is empty, prefer nil.
	if len(bz) == 0 {
		bz = nil
	}

	return bz, nil
}

// Use pbbindings.
func (cdc *Codec) MarshalPBBindings(pbm PBMessager) ([]byte, error) {
	pbo, err := pbm.ToPBMessage(cdc)
	if err != nil {
		return nil, err
	}
	bz, err := proto.Marshal(pbo)
	return bz, err
}

// Panics if error.
func (cdc *Codec) MustMarshal(o interface{}) []byte {
	bz, err := cdc.Marshal(o)
	if err != nil {
		panic(err)
	}
	return bz
}

// MarshalAny encodes the registered object
// wrapped with google.protobuf.Any.
func (cdc *Codec) MarshalAny(o interface{}) ([]byte, error) {
	cdc.doAutoseal()

	// o cannot be nil, otherwise we don't know what type it is.
	if o == nil {
		return nil, errors.New("MarshalAny() requires non-nil argument")
	}

	// Dereference value if pointer.
	var rv, _, _ = maybeDerefValue(reflect.ValueOf(o))
	var rt = rv.Type()

	// rv cannot be an interface.
	if rv.Kind() == reflect.Interface {
		return nil, errors.New("MarshalAny() requires registered concrete type")
	}

	// Make a temporary interface var, to contain the value of o.
	var ivar interface{} = rv.Interface()
	var iinfo *TypeInfo
	iinfo, err := cdc.getTypeInfoWLock(rt)
	if err != nil {
		return nil, err
	}

	// Encode as interface.
	buf := new(bytes.Buffer)
	err = cdc.encodeReflectBinaryInterface(buf, iinfo, reflect.ValueOf(&ivar).Elem(), FieldOptions{}, true)
	if err != nil {
		return nil, err
	}
	bz := buf.Bytes()

	return bz, nil
}

// Panics if error.
func (cdc *Codec) MustMarshalAny(o interface{}) []byte {
	bz, err := cdc.MarshalAny(o)
	if err != nil {
		panic(err)
	}
	return bz
}

//----------------------------------------
// Unmarshal* methods

// Like Unmarshal, but will first decode the byte-length prefix.
// UnmarshalSized will panic if ptr is a nil-pointer.
// Returns an error if not all of bz is consumed.
func (cdc *Codec) UnmarshalSized(bz []byte, ptr interface{}) error {
	if len(bz) == 0 {
		return errors.New("unmarshalSized cannot decode empty bytes")
	}

	// Read byte-length prefix.
	u64, n := binary.Uvarint(bz)
	if n < 0 {
		return errors.Errorf("Error reading msg byte-length prefix: got code %v", n)
	}
	if u64 > uint64(len(bz)-n) {
		return errors.Errorf("Not enough bytes to read in UnmarshalSized, want %v more bytes but only have %v",
			u64, len(bz)-n)
	} else if u64 < uint64(len(bz)-n) {
		return errors.Errorf("Bytes left over in UnmarshalSized, should read %v more bytes but have %v",
			u64, len(bz)-n)
	}
	bz = bz[n:]

	// Decode.
	return cdc.Unmarshal(bz, ptr)
}

// Like Unmarshal, but will first read the byte-length prefix.
// UnmarshalSizedReader will panic if ptr is a nil-pointer.
// If maxSize is 0, there is no limit (not recommended).
func (cdc *Codec) UnmarshalSizedReader(r io.Reader, ptr interface{},
	maxSize int64) (n int64, err error) {
	if maxSize < 0 {
		panic("maxSize cannot be negative.")
	}

	// Read byte-length prefix.
	var l int64
	var buf [binary.MaxVarintLen64]byte
	for i := 0; i < len(buf); i++ {
		_, err = r.Read(buf[i : i+1])
		if err != nil {
			return
		}
		n++
		if buf[i]&0x80 == 0 {
			break
		}
		if n >= maxSize {
			err = errors.Errorf(
				"read overflow, maxSize is %v but uvarint(length-prefix) is itself greater than maxSize",
				maxSize,
			)
		}
	}
	u64, _ := binary.Uvarint(buf[:])
	if err != nil {
		return
	}
	if maxSize > 0 {
		if uint64(maxSize) < u64 {
			err = errors.Errorf("read overflow, maxSize is %v but this amino binary object is %v bytes", maxSize, u64)
			return
		}
		if (maxSize - n) < int64(u64) {
			err = errors.Errorf(
				"read overflow, maxSize is %v but this length-prefixed amino binary object is %v+%v bytes",
				maxSize, n, u64,
			)
			return
		}
	}
	l = int64(u64)
	if l < 0 {
		_ = errors.Errorf( //nolint:errcheck
			"read overflow, this implementation can't read this because, why would anyone have this much data? Hello from 2018",
		)
	}

	// Read that many bytes.
	var bz = make([]byte, l)
	_, err = io.ReadFull(r, bz)
	if err != nil {
		return
	}
	n += l

	// Decode.
	err = cdc.Unmarshal(bz, ptr)
	return n, err
}

// Panics if error.
func (cdc *Codec) MustUnmarshalSized(bz []byte, ptr interface{}) {
	err := cdc.UnmarshalSized(bz, ptr)
	if err != nil {
		panic(err)
	}
}

// Unmarshal will panic if ptr is a nil-pointer.
func (cdc *Codec) Unmarshal(bz []byte, ptr interface{}) error {
	cdc.doAutoseal()

	if cdc.usePBBindings {
		pbm, ok := ptr.(PBMessager)
		if ok {
			return cdc.unmarshalPBBindings(bz, pbm)
		} else {
			// Fall back to using reflection for native primitive types.
		}
	}

	return cdc.unmarshalReflect(bz, ptr)
}

// Use reflection.
func (cdc *Codec) unmarshalReflect(bz []byte, ptr interface{}) error {
	rv := reflect.ValueOf(ptr)
	if rv.Kind() != reflect.Ptr {
		return ErrNoPointer
	}
	rv = rv.Elem()
	rt := rv.Type()
	info, err := cdc.getTypeInfoWLock(rt)
	if err != nil {
		return err
	}

	// See if we need to read the typ3 encoding of an implicit struct.
	//
	// If the dest ptr is an interface, it is assumed that the object is
	// wrapped in a google.protobuf.Any object, so skip this step.
	//
	// See corresponding encoding message in this file, and also
	// binary-decode.
	var bare = true
	var nWrap int
	if !info.IsStructOrUnpacked(FieldOptions{}) &&
		len(bz) > 0 &&
		(rv.Kind() != reflect.Interface) {
		var (
			fnum      uint32
			typ       Typ3
			nFnumTyp3 int
		)
		fnum, typ, nFnumTyp3, err = decodeFieldNumberAndTyp3(bz)
		if err != nil {
			return errors.Wrap(err, "could not decode field number and type")
		}
		if fnum != 1 {
			return fmt.Errorf("expected field number: 1; got: %v", fnum)
		}
		typWanted := info.GetTyp3(FieldOptions{})
		if typ != typWanted {
			return fmt.Errorf("expected field type %v for # %v of %v, got %v",
				typWanted, fnum, info.Type, typ)
		}

		slide(&bz, &nWrap, nFnumTyp3)
		// "bare" is ignored when primitive, byteslice, bytearray.
		// When typ3 != ByteLength, then typ3 is one of Typ3Varint, Typ38Byte,
		// Typ34Byte; and they are all primitive.
		bare = false
	}

	// Decode contents into rv.
	n, err := cdc.decodeReflectBinary(bz, info, rv, FieldOptions{BinFieldNum: 1}, bare, 0)
	if err != nil {
		return fmt.Errorf(
			"unmarshal to %v failed after %d bytes (%v): %X",
			info.Type,
			n+nWrap,
			err,
			bz,
		)
	}
	if n != len(bz) {
		return fmt.Errorf(
			"unmarshal to %v didn't read all bytes. Expected to read %v, only read %v: %X",
			info.Type,
			len(bz),
			n+nWrap,
			bz,
		)
	}

	return nil
}

// Use pbbindings.
func (cdc *Codec) unmarshalPBBindings(bz []byte, pbm PBMessager) error {
	pbo := pbm.EmptyPBMessage(cdc)
	err := proto.Unmarshal(bz, pbo)
	if err != nil {
		return err
	}
	err = pbm.FromPBMessage(cdc, pbo)
	if err != nil {
		return err
	}
	return nil
}

// Panics if error.
func (cdc *Codec) MustUnmarshal(bz []byte, ptr interface{}) {
	err := cdc.Unmarshal(bz, ptr)
	if err != nil {
		panic(err)
	}
}

// UnmarshalAny decodes the registered object
// from the Any fields.
func (cdc *Codec) UnmarshalAny(typeURL string, value []byte, ptr interface{}) (err error) {
	cdc.doAutoseal()

	rv := reflect.ValueOf(ptr)
	if rv.Kind() != reflect.Ptr {
		return ErrNoPointer
	}
	rv = rv.Elem()
	_, err = cdc.decodeReflectBinaryAny(typeURL, value, rv, FieldOptions{})
	return
}

func (cdc *Codec) MustUnmarshalAny(typeURL string, value []byte, ptr interface{}) {
	err := cdc.UnmarshalAny(typeURL, value, ptr)
	if err != nil {
		panic(err)
	}
	return
}

func (cdc *Codec) MarshalJSON(o interface{}) ([]byte, error) {
	cdc.doAutoseal()

	rv := reflect.ValueOf(o)
	if rv.Kind() == reflect.Invalid {
		return []byte("null"), nil
	}
	rt := rv.Type()
	w := new(bytes.Buffer)
	info, err := cdc.getTypeInfoWLock(rt)
	if err != nil {
		return nil, err
	}
	if err = cdc.encodeReflectJSON(w, info, rv, FieldOptions{}); err != nil {
		return nil, err
	}
	return w.Bytes(), nil
}

func (cdc *Codec) MarshalJSONAny(o interface{}) ([]byte, error) {
	// o cannot be nil, otherwise we don't know what type it is.
	if o == nil {
		return nil, errors.New("MarshalJSONAny() requires non-nil argument")
	}

	// Dereference value if pointer.
	var rv = reflect.ValueOf(o)
	if rv.Kind() == reflect.Ptr {
		rv = rv.Elem()
	}
	var rt = rv.Type()

	// rv cannot be an interface.
	if rv.Kind() == reflect.Interface {
		return nil, errors.New("MarshalJSONAny() requires registered concrete type")
	}

	// Make a temporary interface var, to contain the value of o.
	var ivar interface{} = rv.Interface()
	var iinfo *TypeInfo
	iinfo, err := cdc.getTypeInfoWLock(rt)
	if err != nil {
		return nil, err
	}

	// Encode as interface.
	buf := new(bytes.Buffer)
	err = cdc.encodeReflectJSONInterface(buf, iinfo, reflect.ValueOf(&ivar).Elem(), FieldOptions{})
	if err != nil {
		return nil, err
	}
	bz := buf.Bytes()

	return bz, nil
}

// MustMarshalJSON panics if an error occurs. Besides tha behaves exactly like MarshalJSON.
func (cdc *Codec) MustMarshalJSON(o interface{}) []byte {
	bz, err := cdc.MarshalJSON(o)
	if err != nil {
		panic(err)
	}
	return bz
}

func (cdc *Codec) UnmarshalJSON(bz []byte, ptr interface{}) error {
	cdc.doAutoseal()
	if len(bz) == 0 {
		return errors.New("cannot decode empty bytes")
	}

	rv := reflect.ValueOf(ptr)
	if rv.Kind() != reflect.Ptr {
		return errors.New("expected a pointer")
	}
	rv = rv.Elem()
	rt := rv.Type()
	info, err := cdc.getTypeInfoWLock(rt)
	if err != nil {
		return err
	}
	return cdc.decodeReflectJSON(bz, info, rv, FieldOptions{})
}

// MustUnmarshalJSON panics if an error occurs. Besides tha behaves exactly like UnmarshalJSON.
func (cdc *Codec) MustUnmarshalJSON(bz []byte, ptr interface{}) {
	if err := cdc.UnmarshalJSON(bz, ptr); err != nil {
		panic(err)
	}
}

// MarshalJSONIndent calls json.Indent on the output of cdc.MarshalJSON
// using the given prefix and indent string.
func (cdc *Codec) MarshalJSONIndent(o interface{}, prefix, indent string) ([]byte, error) {
	bz, err := cdc.MarshalJSON(o)
	if err != nil {
		return nil, err
	}
	var out bytes.Buffer
	err = json.Indent(&out, bz, prefix, indent)
	if err != nil {
		return nil, err
	}
	return out.Bytes(), nil
}

//----------------------------------------
// Other

// NOTE: do not modify the result.
func RegisterPackage(pi *pkg.Package) *Package {
	gcdc.RegisterPackage(pi)
	return pi
}

func NewPackage(gopkg string, p3pkg string, dirname string) *Package {
	return pkg.NewPackage(gopkg, p3pkg, dirname)
}

// NOTE: duplicated in pkg/pkg.go
func GetCallersDirname() string {
	var dirname = "" // derive from caller.
	_, filename, _, ok := runtime.Caller(1)
	if !ok {
		panic("could not get caller to derive caller's package directory")
	}
	dirname = path.Dir(filename)
	if filename == "" || dirname == "" {
		panic("could not derive caller's package directory")
	}
	return dirname
}

//----------------------------------------
// Object

// All concrete types must implement the Object interface for genproto
// bindings.  They are generated automatically by genproto/bindings.go
type Object interface {
	GetTypeURL() string
}

// TODO: this does need the cdc receiver,
// as it should also work for non-pbbindings-optimized types.
// Returns the default type url for the given concrete type.
// NOTE: It must be fast, as it is used in pbbindings.
// XXX Unstable API.
func (cdc *Codec) GetTypeURL(o interface{}) string {
	if obj, ok := o.(Object); ok {
		return obj.GetTypeURL()
	}
	switch o.(type) {
	case time.Time, *time.Time, *timestamppb.Timestamp:
		return "/google.protobuf.Timestamp"
	case time.Duration, *time.Duration, *durationpb.Duration:
		return "/google.protobuf.Duration"
	}
	rv := reflect.ValueOf(o)
	// Wellknown override.
	// TODO: move this to wellknown somehow.
	switch rv.Kind() {
	case reflect.String:
		return "/google.protobuf.StringValue"
	case reflect.Int64, reflect.Int:
		return "/google.protobuf.Int64Value"
	case reflect.Int32:
		return "/google.protobuf.Int32Value"
	case reflect.Int16:
		return "/google.protobuf.Int32Value"
	case reflect.Int8:
		return "/google.protobuf.Int32Value"
	case reflect.Uint64, reflect.Uint:
		return "/google.protobuf.UInt64Value"
	case reflect.Uint32:
		return "/google.protobuf.UInt32Value"
	case reflect.Uint16:
		return "/google.protobuf.UInt32Value"
	case reflect.Uint8:
		return "/google.protobuf.UInt32Value"
	case reflect.Bool:
		return "/google.protobuf.BoolValue"
	case reflect.Array:
		if rv.Type().Elem().Kind() == reflect.Uint8 {
			return "/google.protobuf.BytesValue"
		} else {
			panic("not yet supported")
		}
	case reflect.Slice:
		if rv.Type().Elem().Kind() == reflect.Uint8 {
			return "/google.protobuf.BytesValue"
		} else {
			panic("not yet supported")
		}
	}
	// Doesn't have .GetTypeURL() and isn't well known.
	// Do the slow thing (not relevant if pbbindings exists).
	info, err := cdc.GetTypeInfo(rv.Type())
	if err != nil {
		panic(err)
	}
	if info.TypeURL == "" {
		panic("not yet supported")
	}
	return info.TypeURL
}

//----------------------------------------

// Methods generated by genproto/bindings.go for faster encoding.
type PBMessager interface {
	ToPBMessage(*Codec) (proto.Message, error)
	EmptyPBMessage(*Codec) proto.Message
	FromPBMessage(*Codec, proto.Message) error
}
